R/crunch-data-frame.R

```diff
@@ -0,0 +1,292 @@
+    }
+    
+    return(out)
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`[<-.CrunchDataFrame` <- function (x, i, j, value) {
+    j <- grab_col_names(x, j)
+    row_inds <- grab_row_ind(x, i)
+    
+    # TODO: check value length
+    
+    # split value string by the length of row indicators
+    values <- split(value, ceiling(seq_along(value)/length(row_inds)))
+    for (i in seq_along(j)) {
+        col_name <- j[i]
+        x <- set_CDF_var(col_name = col_name, row_inds = row_inds,
+                         cdf = x, value = values[[i]])
+    }
+    return(x)
+}
+
+grab_col_names <- function (x, j) {
+    if (missing(j)) {
+        # if there's no j, grab all columns
+        j <- names(x)
+    } else if (is.logical(j)) {
+        if (length(j) != ncol(x)) {
+            halt("when using a logical to subset columns, the logical vector ",
+                 "must have the same length as the number of columns.")
+        }
+        j <- names(x)[j]
+    }
+    return(j)
+}
+
+grab_row_ind <- function (x, i) {
+    if (missing(i)) {
+        # if there is no row specification, use all rows.
+        row_inds <- seq_len(nrow(x))
+    } else {
+        if (is.logical(i)) {
+            if (length(i) != nrow(x)) {
+                halt("when using a logical to subset rows, the logical vector ",
+                     "must have the same length as the number of rows")
+            }
+        } else if (!is.numeric(i)) {
+            halt("row subsetting must be done with either numeric or a logical.")
+        }
+        row_inds <- i
+    }
+    return(row_inds)
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`[[.CrunchDataFrame` <- function (x, i) {
+    col_name <- col_name_from_index(x, i)
+    
+    return(get_CDF_var(col_name, x, mode = attr(x, "mode")))
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`[[<-.CrunchDataFrame` <- function (x, i, value) {
+    col_name <- col_name_from_index(x, i)
+    
+    x <- set_CDF_var(col_name = col_name, cdf = x, value = value)
+    return(x)
+}
+
+col_name_from_index <- function(x, i) {
+    if (is.character(i)) {
+        col_name <- i
+        # check if any don't exist
+    } else if (is.numeric(i)) {
+        col_name <- names(x)[i]
+    } else {
+        halt("column subsetting must be done with either numeric or ",
+             "a character")
+    }
+    return(col_name)
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`$.CrunchDataFrame` <- function (x, i) {
+    return(get_CDF_var(i, x, mode = attr(x, "mode")))
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`$<-.CrunchDataFrame` <- function (x, i, value) {
+    x <- set_CDF_var(col_name = i, cdf = x, value = value)
+    return(x)
+}
+
+get_CDF_var <- function (col_name, cdf, ...) {
+    if (!is(cdf, "CrunchDataFrame")) { 
+        halt("The cdf argument must be a CrunchDataFrame, got ", class(cdf), " instead.")
+    }
+    if (!col_name %in% names(cdf)) {
+        halt("The variable ", dQuote(col_name), " is not found in the CrunchDataFrame.")
```

Comment:
Is this check required? How cryptic is the error returned by `get` if you let it handle whether the variable exists?

Reply:
It's not *super* cryptic, but it's not ideal since it references objects rather than columns or variables: `Error in get(col_name, crdf) : object 'fladskj' not found`